{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 52%|██████████████████████████████████████▏                                  | 52343/100000 [00:16<00:15, 3113.25it/s]\n",
      "  0%|                                                                                          | 0/500 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████████████████████████████████████████████████████████████████████████| 500/500 [1:15:50<00:00,  9.10s/it]\n"
     ]
    }
   ],
   "source": [
    "from rnn import Encoder\n",
    "from preprocess import proc\n",
    "import tensorflow as tf\n",
    "from tqdm import tqdm\n",
    "import pandas as pd\n",
    "\n",
    "encoder = Encoder(300, 300, 100, 1)\n",
    "encoder.load_weights(\"encoder.weights\")\n",
    "\n",
    "w2v_dict = dict()\n",
    "f = open(\"./glove.840B.300d.txt\", \"rt\", encoding='utf-8')\n",
    "for i in tqdm(range(100000)):\n",
    "    line = f.readline()\n",
    "    els = line.split()\n",
    "    token = els[0]\n",
    "    try:\n",
    "        value = tf.convert_to_tensor(list(map(float, els[1:])), dtype=tf.float32)\n",
    "    except:\n",
    "        break\n",
    "    w2v_dict[token] = value\n",
    "f.close()\n",
    "\n",
    "def comment_text_to_vec(comment_text):\n",
    "    p = proc(comment_text, [1, 2])\n",
    "    tweet = []\n",
    "    for word in p:\n",
    "        if word in w2v_dict:\n",
    "            tweet.append(w2v_dict[word])\n",
    "    if len(tweet) == 0:\n",
    "        return None\n",
    "    tweet = tf.stack(tweet)\n",
    "    tweet = tf.expand_dims(tweet, axis=1)\n",
    "    tweet = tf.expand_dims(tweet, axis=3)\n",
    "    enc_hidden = encoder.initialize_hidden_state()\n",
    "    for word in tweet:\n",
    "        out_t, enc_hidden = encoder(word, enc_hidden)\n",
    "    return enc_hidden[0].numpy()\n",
    "\n",
    "\n",
    "def cross_validation(model, X, y, scorer, cv=5):\n",
    "    return cross_validate(estimator=model,\n",
    "                          X=X,\n",
    "                          y=y,\n",
    "                          cv=cv,\n",
    "                          scoring=scorer,\n",
    "                          return_train_score=True)\n",
    "\n",
    "\n",
    "train = pd.read_csv(\"./train.csv\")[['target', 'comment_text']][:10000]\n",
    "train['toxic'] = 1 * (train['target'] > 0.5)\n",
    "\n",
    "print(\"Training\")\n",
    "X, y = [], []\n",
    "for Xrow, yrow in tqdm(list(zip(train['comment_text'], train[\"toxic\"]))[:500]):\n",
    "    v = comment_text_to_vec(Xrow)\n",
    "    if v is not None:\n",
    "        X.append(v)\n",
    "        y.append(yrow)\n",
    "\n",
    "import numpy as np\n",
    "X = np.stack(X)\n",
    "y = np.array(y)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 1312, in fit\n",
      "    raise ValueError(\"l1_ratio must be between 0 and 1;\"\n",
      "ValueError: l1_ratio must be between 0 and 1; got (l1_ratio=None)\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 1312, in fit\n",
      "    raise ValueError(\"l1_ratio must be between 0 and 1;\"\n",
      "ValueError: l1_ratio must be between 0 and 1; got (l1_ratio=None)\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 1407, in fit\n",
      "    fold_coefs_ = Parallel(n_jobs=self.n_jobs, verbose=self.verbose,\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 1029, in __call__\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 847, in dispatch_one_batch\n",
      "    self._dispatch(tasks)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 765, in _dispatch\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\", line 208, in apply_async\n",
      "    result = ImmediateResult(func)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\", line 572, in __init__\n",
      "    self.results = batch()\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 252, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 252, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 762, in _logistic_regression_path\n",
      "    n_iter_i = _check_optimize_result(\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\utils\\optimize.py\", line 243, in _check_optimize_result\n",
      "    ).format(solver, result.status, result.message.decode(\"latin1\"))\n",
      "AttributeError: 'str' object has no attribute 'decode'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 1407, in fit\n",
      "    fold_coefs_ = Parallel(n_jobs=self.n_jobs, verbose=self.verbose,\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 1029, in __call__\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 847, in dispatch_one_batch\n",
      "    self._dispatch(tasks)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 765, in _dispatch\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\", line 208, in apply_async\n",
      "    result = ImmediateResult(func)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\", line 572, in __init__\n",
      "    self.results = batch()\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 252, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 252, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 762, in _logistic_regression_path\n",
      "    n_iter_i = _check_optimize_result(\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\utils\\optimize.py\", line 243, in _check_optimize_result\n",
      "    ).format(solver, result.status, result.message.decode(\"latin1\"))\n",
      "AttributeError: 'str' object has no attribute 'decode'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 1407, in fit\n",
      "    fold_coefs_ = Parallel(n_jobs=self.n_jobs, verbose=self.verbose,\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 1029, in __call__\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 847, in dispatch_one_batch\n",
      "    self._dispatch(tasks)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 765, in _dispatch\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\", line 208, in apply_async\n",
      "    result = ImmediateResult(func)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\", line 572, in __init__\n",
      "    self.results = batch()\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 252, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 252, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 762, in _logistic_regression_path\n",
      "    n_iter_i = _check_optimize_result(\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\utils\\optimize.py\", line 243, in _check_optimize_result\n",
      "    ).format(solver, result.status, result.message.decode(\"latin1\"))\n",
      "AttributeError: 'str' object has no attribute 'decode'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 1407, in fit\n",
      "    fold_coefs_ = Parallel(n_jobs=self.n_jobs, verbose=self.verbose,\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 1029, in __call__\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 847, in dispatch_one_batch\n",
      "    self._dispatch(tasks)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 765, in _dispatch\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\", line 208, in apply_async\n",
      "    result = ImmediateResult(func)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\", line 572, in __init__\n",
      "    self.results = batch()\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 252, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 252, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 762, in _logistic_regression_path\n",
      "    n_iter_i = _check_optimize_result(\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\utils\\optimize.py\", line 243, in _check_optimize_result\n",
      "    ).format(solver, result.status, result.message.decode(\"latin1\"))\n",
      "AttributeError: 'str' object has no attribute 'decode'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 1407, in fit\n",
      "    fold_coefs_ = Parallel(n_jobs=self.n_jobs, verbose=self.verbose,\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 1029, in __call__\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 847, in dispatch_one_batch\n",
      "    self._dispatch(tasks)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 765, in _dispatch\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\", line 208, in apply_async\n",
      "    result = ImmediateResult(func)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\", line 572, in __init__\n",
      "    self.results = batch()\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 252, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 252, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 762, in _logistic_regression_path\n",
      "    n_iter_i = _check_optimize_result(\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\utils\\optimize.py\", line 243, in _check_optimize_result\n",
      "    ).format(solver, result.status, result.message.decode(\"latin1\"))\n",
      "AttributeError: 'str' object has no attribute 'decode'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 1407, in fit\n",
      "    fold_coefs_ = Parallel(n_jobs=self.n_jobs, verbose=self.verbose,\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 1029, in __call__\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 847, in dispatch_one_batch\n",
      "    self._dispatch(tasks)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 765, in _dispatch\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\", line 208, in apply_async\n",
      "    result = ImmediateResult(func)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\", line 572, in __init__\n",
      "    self.results = batch()\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 252, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 252, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 762, in _logistic_regression_path\n",
      "    n_iter_i = _check_optimize_result(\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\utils\\optimize.py\", line 243, in _check_optimize_result\n",
      "    ).format(solver, result.status, result.message.decode(\"latin1\"))\n",
      "AttributeError: 'str' object has no attribute 'decode'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 1407, in fit\n",
      "    fold_coefs_ = Parallel(n_jobs=self.n_jobs, verbose=self.verbose,\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 1029, in __call__\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 847, in dispatch_one_batch\n",
      "    self._dispatch(tasks)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 765, in _dispatch\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\", line 208, in apply_async\n",
      "    result = ImmediateResult(func)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\", line 572, in __init__\n",
      "    self.results = batch()\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 252, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 252, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 762, in _logistic_regression_path\n",
      "    n_iter_i = _check_optimize_result(\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\utils\\optimize.py\", line 243, in _check_optimize_result\n",
      "    ).format(solver, result.status, result.message.decode(\"latin1\"))\n",
      "AttributeError: 'str' object has no attribute 'decode'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 1407, in fit\n",
      "    fold_coefs_ = Parallel(n_jobs=self.n_jobs, verbose=self.verbose,\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 1029, in __call__\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 847, in dispatch_one_batch\n",
      "    self._dispatch(tasks)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 765, in _dispatch\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\", line 208, in apply_async\n",
      "    result = ImmediateResult(func)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\", line 572, in __init__\n",
      "    self.results = batch()\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 252, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 252, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 762, in _logistic_regression_path\n",
      "    n_iter_i = _check_optimize_result(\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\utils\\optimize.py\", line 243, in _check_optimize_result\n",
      "    ).format(solver, result.status, result.message.decode(\"latin1\"))\n",
      "AttributeError: 'str' object has no attribute 'decode'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.10210254700791385 l2 False\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 1407, in fit\n",
      "    fold_coefs_ = Parallel(n_jobs=self.n_jobs, verbose=self.verbose,\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 1029, in __call__\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 847, in dispatch_one_batch\n",
      "    self._dispatch(tasks)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 765, in _dispatch\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\", line 208, in apply_async\n",
      "    result = ImmediateResult(func)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\", line 572, in __init__\n",
      "    self.results = batch()\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 252, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 252, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 762, in _logistic_regression_path\n",
      "    n_iter_i = _check_optimize_result(\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\utils\\optimize.py\", line 243, in _check_optimize_result\n",
      "    ).format(solver, result.status, result.message.decode(\"latin1\"))\n",
      "AttributeError: 'str' object has no attribute 'decode'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n",
      "C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 531, in _fit_and_score\n",
      "    estimator.fit(X_train, y_train, **fit_params)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 1407, in fit\n",
      "    fold_coefs_ = Parallel(n_jobs=self.n_jobs, verbose=self.verbose,\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 1029, in __call__\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 847, in dispatch_one_batch\n",
      "    self._dispatch(tasks)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 765, in _dispatch\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\", line 208, in apply_async\n",
      "    result = ImmediateResult(func)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\", line 572, in __init__\n",
      "    self.results = batch()\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 252, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\joblib\\parallel.py\", line 252, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py\", line 762, in _logistic_regression_path\n",
      "    n_iter_i = _check_optimize_result(\n",
      "  File \"C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\utils\\optimize.py\", line 243, in _check_optimize_result\n",
      "    ).format(solver, result.status, result.message.decode(\"latin1\"))\n",
      "AttributeError: 'str' object has no attribute 'decode'\n",
      "\n",
      "  warnings.warn(\"Estimator fit failed. The score on this train-test\"\n"
     ]
    }
   ],
   "source": [
    "\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.model_selection import cross_validate\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "scaler = StandardScaler()\n",
    "Xs = scaler.fit_transform(X)\n",
    "\n",
    "best = 0\n",
    "pen_best = 0\n",
    "inter_best = 0\n",
    "for penalty in ('l1', 'l2', 'elasticnet', 'none'):\n",
    "    solver = 'lbfgs'\n",
    "    if penalty == 'elasticnet':\n",
    "        solver = 'saga'\n",
    "    if penalty == 'l1':\n",
    "        solver = 'liblinear'\n",
    "    for intercept in (True, False):\n",
    "        m = LogisticRegression(penalty = penalty, fit_intercept = intercept, solver = solver)\n",
    "        res = cross_validation(m, Xs, y, ['f1'], 5)\n",
    "        if res['test_f1'].mean() > best:\n",
    "            best = res['test_f1'].mean()\n",
    "            pen_best = penalty\n",
    "            inter_best = intercept\n",
    "\n",
    "print(best, pen_best, inter_best)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\hp\\anaconda3\\lib\\site-packages\\sklearn\\metrics\\_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.926829268292683\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.93      1.00      0.96       114\n",
      "           1       0.00      0.00      0.00         9\n",
      "\n",
      "    accuracy                           0.93       123\n",
      "   macro avg       0.46      0.50      0.48       123\n",
      "weighted avg       0.86      0.93      0.89       123\n",
      "\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAATgAAAEGCAYAAADxD4m3AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAWqUlEQVR4nO3debRdZXnH8e8vN4EQIIGYwUsIEjSNMhSxEUFaG8QKSpfBLlEUaVQqakGs1Va0VloHSpdDtQpqBDQOgEHRxKoQVxCRViEhIDIUQxlCSEImmSHDvU//2PvGQ7g52fvk7LuH+/ustVfO3uecdz/3hjy8434VEZiZNdGIsgMwMyuKE5yZNZYTnJk1lhOcmTWWE5yZNdbIsgNoNWF8Txw4dVTZYVgOv7t1TNkhWA5P8wSbY5N2pYzjj90zNmzsy/TZm27ddHVEnLAr99sVlUpwB04dxY1XTy07DMvh+P1eXHYIlsMNsXiXy9iwsY8brz4g02d7epdP2OUb7oJKJTgzq74A+ukvO4xMnODMLJcg2BLZmqhlc4Izs9xcgzOzRgqCvpos8XSCM7Pc+nGCM7MGCqDPCc7Mmso1ODNrpAC2uA/OzJooCDdRzayhAvrqkd+c4Mwsn2QlQz04wZlZTqKPXVqvP2Sc4Mwsl2SQwQnOzBoomQfnBGdmDdXvGpyZNZFrcGbWWIHoq8luB05wZpabm6hm1kiB2Bw9ZYeRiROcmeWSTPR1E9XMGsqDDGbWSBGiL+pRg6tHlGZWKf0o07Ezki6RtFbSbS3Xxkv6maTl6Z/7trz3YUl3S7pL0vE7K98JzsxySQYZRmY6MvgGsP3G0OcAiyNiOrA4PUfSwcApwCHpdy6U1Ha0wwnOzHIZGGTIcuy0rIjrgI3bXZ4NzEtfzwNOarl+eURsioh7gbuBI9uV7z44M8utL/s8uAmSlracz42IuTv5zuSIWA0QEaslTUqvTwF+3fK5lem1HXKCM7Nccq5kWB8RM7t068GyattHbzrBmVlu/cWOoj4kqTetvfUCa9PrK4GpLZ/bH1jVriD3wZlZLsli+xGZjg4tBOakr+cAC1qunyJpd0nTgOnAje0Kcg3OzHIJxJYuLdWSdBkwi6SvbiVwLnA+MF/S6cAK4GSAiLhd0nzgDmArcGZE9LUr3wnOzHKJoGsTfSPizTt467gdfP5TwKeylu8EZ2Y5ZZvEWwVOcGaWS9C9GlzRnODMLDc/8NLMGimQH3hpZs2UbBtYj9RRjyjNrEK88bOZNVRQ+EqGrnGCM7PcXIMzs0aKkGtwZtZMySCDd9Uys0aqz54MTnBmlksyyOA+ODNrKK9kMLNG8koGM2s072xvZo0UAVv6neDMrIGSJqoTnJk1VF1WMtQjDVfYZ98/lTcedghnHDtj27XrfjSOd86awQlTDud3v9njWd9Zu3IUs19wGFd8eeJQhmoZzJz1KBf98n/5+n/fyRvPeqjscCppYJpIlqNshSY4SSdIukvS3ZLOKfJeZXn1mzbyqe/c84xrB77waT520X0cdtQTg37nK/8yhZe+8rGhCM9yGDEiOPO8B/noqdN456wZHDv7YQ6Y/nTZYVVQ0kTNcpStsCaqpB7gAuAvSPYzXCJpYUTcUdQ9y3DYUU+w5oHdnnHtgOmbdvj5//npOHoP2MzoMf1Fh2Y5zTjiSVbdtxtrVuwOwLUL9uHo4x9hxfLRJUdWPXXZk6HIFHskcHdE3BMRm4HLgdkF3q/ynn5yBPMvnMRbP7Cm7FBsEM957hbWrfrD/6zWrx7FhN4tJUZUTckoak+mo2xFJrgpwAMt5yvTa88g6QxJSyUtXbeh7RaHtffNTz+X179zHXvs6dpbFWmQSknE0MdRdQMTfevQB1fkKOpgP92z/nOJiLnAXICZh49u9H9O/3vzGK7/8T5c/Mn9ePzRHjQi2G33YPY71pcdmpHU2Cbut3nb+YTeLWxYM6rEiKqrLk3UIhPcSmBqy/n+wKoC71d5n/vh3dtef+szz2X0nn1ObhVy1y1jmDJtM5OnbmLDmlHMmv0w55/5vLLDqhwvtk8sAaZLmgY8CJwCvKXA+5Xi397zPG791V48snEkp/7JwZz2gTXsvW8fF350Co9sGMk/n3YQzz/kKc677J6dF2al6u8TF/zTFM679B5G9MCiy8dz/+88wDCYKoyQZlFYgouIrZLOAq4GeoBLIuL2ou5Xlg9/+f5Brx/zmkfafu+0D3qgoYqWXDOWJdeMLTuMSosQW4d7ggOIiJ8APynyHmY29NxENbNGqlMfXD3qmWZWKd2aJiLp/ZJul3SbpMskjZY0XtLPJC1P/9y30zid4Mwsl27Ng5M0BTgbmBkRh5L01Z8CnAMsjojpwOL0vCNOcGaWWz/KdGQwEthD0khgDMlUstnAvPT9ecBJncbpPjgzyyUCtmZ/4OUESUtbzuemk/uJiAclfQZYATwFLIqIRZImR8Tq9DOrJU3qNFYnODPLLccgw/qImDnYG2nf2mxgGvAwcIWkt3YnwoQTnJnl0sVNZ14F3BsR6wAkXQm8HHhIUm9ae+sF1nZ6A/fBmVluEcp07MQK4ChJYyQJOA64E1gIzEk/MwdY0GmcrsGZWW7dWGwfETdI+h6wDNgK3Ezy4I29gPmSTidJgid3eg8nODPLJaJ7E30j4lzg3O0ubyKpze0yJzgzy0n0edtAM2uqDP1rleAEZ2a51GktqhOcmeUT9XmUuxOcmeXmR5abWSOFBxnMrMncRDWzxvIoqpk1UoQTnJk1mKeJmFljuQ/OzBopEP0eRTWzpqpJBc4Jzsxy8iCDmTVaTapwTnBmllvta3CSvkibPB0RZxcSkZlVWgD9/TVPcMDSNu+Z2XAVQN1rcBExr/Vc0p4R8UTxIZlZ1dVlHtxOJ7NIOlrSHSS73SDpcEkXFh6ZmVVXZDxKlmW23ueB44ENABHxG+AVRQZlZlWWbcvAKgxEZBpFjYgHkm0Lt+krJhwzq4UK1M6yyJLgHpD0ciAk7QacTdpcNbNhKCBqMoqapYn6buBMYArwIPDi9NzMhi1lPMq10xpcRKwHTh2CWMysLmrSRM0yinqQpB9JWidpraQFkg4aiuDMrKIaNIp6KTAf6AX2A64ALisyKDOrsIGJvlmOkmVJcIqIb0XE1vT4NpXIzWZWlohsR9narUUdn778uaRzgMtJEtubgB8PQWxmVlU1GUVtN8hwE0lCG/hJ3tXyXgCfKCooM6s2dal2Jmkf4CLgUJK88g7gLuC7wIHAfcAbI+L3nZTfbi3qtE4KNLOG6+4AwheAqyLiDek82zHAR4DFEXF+2no8B/hQJ4VnWskg6VDgYGD0wLWI+GYnNzSzuuvOAIKksSTLPt8GEBGbgc2SZgOz0o/NA66lqAQn6dz0ZgcDPwFeA1wPOMGZDVfZa3ATJLU+em1uRMxNXx8ErAO+Lulwkm6x9wGTI2I1QESsljSp0zCz1ODeABwO3BwRb5c0maTNbGbDVX/mT66PiJk7eG8k8BLgvRFxg6QvkDRHuybLNJGnIqIf2JpWKdeSZF4zG466Nw9uJbAyIm5Iz79HkvAektQLkP65ttNQsyS4pelIx9dIqpDLgBs7vaGZ1Z8i29FORKwheZjHjPTSccAdwEJgTnptDrCg0zizrEX92/TlVyRdBYyNiFs7vaGZNUD3RlHfC3wnHUG9B3g7ScVrvqTTgRXAyZ0W3m6i70vavRcRyzq9qZkZQETcAgzWR3dcN8pvV4P7bJv3AnhlNwJotfzOcZx45IndLtYK9WDZAVgJujXRt2jtJvoeO5SBmFlNBI1YqmVmNri61+DMzHak9k1UM7MdqkmCy/JEX0l6q6SPpecHSDqy+NDMrLIa9ETfC4GjgTen548BFxQWkZlVWtZJvlVoxmZpor4sIl4i6WaAiPh9OinPzIarBo2ibpHUQ1rhlDSRPEttzaxxqlA7yyJLE/U/gR8AkyR9iuRRSecVGpWZVVtN+uCyrEX9jqSbSJZOCDgpIryzvdlwVZH+tSyyPPDyAOBJ4Eet1yJiRZGBmVmFNSXBkeygNbD5zGhgGsmmEIcUGJeZVZhq0gufpYl6WOt5+pSRd+3g42ZmlZF7JUNELJP00iKCMbOaaEoTVdLft5yOIHmk8LrCIjKzamvSIAOwd8vrrSR9ct8vJhwzq4UmJLh0gu9eEfEPQxSPmdVB3ROcpJERsbXdo8vNbPgRzRhFvZGkv+0WSQuBK4AnBt6MiCsLjs3MqqhhfXDjgQ0kezAMzIcLwAnObLhqQIKblI6g3sYfEtuAmvx4ZlaImmSAdgmuB9iLZya2ATX58cysCE1ooq6OiI8PWSRmVh8NSHD1eKKdmQ2taMYoald2ljazBqp7DS4iNg5lIGZWH03ogzMzG5wTnJk1UkUeR55Flj0ZzMy2Ed3dNlBSj6SbJf1Xej5e0s8kLU//3LfTWJ3gzCy3Lu+L+j6gdZ+Xc4DFETEdWJyed8QJzszy69KuWpL2B04ELmq5PBuYl76eB5zUaZhOcGaWX/e2Dfw88I88c6/lyRGxGiD9c1KnYTrBmVk+GZunaRN1gqSlLccZA8VI+ktgbUTcVFSoHkU1s/yy96+tj4iZO3jvGOB1kl5LsmPfWEnfBh6S1BsRqyX1Ams7DdM1ODPLTf3ZjnYi4sMRsX9EHAicAlwTEW8FFgJz0o/NARZ0GqdrcGaWW8ErGc4H5ks6HVgBnNxpQU5wZpZPARN9I+Ja4Nr09Qa6tBbeCc7M8qvJSgYnODPLZWAlQx04wZlZbuqvR4ZzgjOzfGq02N4JzsxycxPVzJrLCc7Mmso1ODNrLic4M2ukhuyqZWb2LJ4HZ2bNFvXIcE5wZpaba3DG6950L8ef9AASXP3DqSy4fFrZIdlOzJz1KO/+xCp6RgQ/vWw88780ueyQqqdGE30Lex6cpEskrZV0W1H3qLLnHfQYx5/0AH//tmM469Q/5cg/Xct+U58oOyxrY8SI4MzzHuSjp07jnbNmcOzshzlg+tNlh1VJ3Xge3FAo8oGX3wBOKLD8Sps67XHuum0fNm3qob9vBL9dNp6jZ60pOyxrY8YRT7Lqvt1Ys2J3tm4ZwbUL9uHo4x8pO6xKGvYJLiKuAzYWVX7V3f9/e3PoERvZe9xmdt+9j5nHrGPiZNcGquw5z93CulW7bTtfv3oUE3q3lBhRRQXJIEOWo2Sl98Glm1CcATC6Z++So+meB+7bi+998/l88os38vRTPdy7fG/6+lR2WNaGBvnrqcC/0UryIENGETEXmAswbrfJNfm1ZbNo4VQWLZwKwF+/5y42rB1dckTWzvrVo5i43+Zt5xN6t7BhzagSI6qwmvxL9aYzBRq37yYAJk5+ipcfu4ZfLNqv5IisnbtuGcOUaZuZPHUTI0f1M2v2w/x60biyw6qcgYm+XdzZvjCl1+Ca7CP/voyxY7ewtU98+dOH8Phjrg1UWX+fuOCfpnDepfcwogcWXT6e+3/nWvezRPiBl5IuA2aRbPy6Ejg3Ii4u6n5V9KEzji47BMtpyTVjWXLN2LLDqL565LfiElxEvLmoss2sXFVofmbhJqqZ5RPAcG+imlmD1SO/OcGZWX5uoppZYw37UVQza6gaPU3ECc7Mckkm+tYjwznBmVl+FXhSSBZeqmVmuSki09G2DGmqpJ9LulPS7ZLel14fL+lnkpanf+7baZxOcGaWT+Q42tsKfCAiXgQcBZwp6WDgHGBxREwHFqfnHXGCM7OckrWoWY62pUSsjohl6evHgDuBKcBsYF76sXnASZ1G6j44M8uvy4MMkg4EjgBuACZHxOrkNrFa0qROy3WCM7N88m38PEHS0pbzuekzILeRtBfwfeDvIuJRDfbk0Q45wZlZftlrcOsjYuaO3pQ0iiS5fScirkwvPySpN6299QJrOw3TfXBmll8XBhmUVNUuBu6MiM+1vLUQmJO+ngMs6DRM1+DMLDf1d2Ui3DHAacBvJd2SXvsIcD4wX9LpwArg5E5v4ARnZvkEXZnoGxHXkyyMGMxxu34HJzgzy0nsfBJvVTjBmVl+TnBm1lhOcGbWSF3qgxsKTnBmlluXRlEL5wRnZjmFm6hm1lCBE5yZNVg9WqhOcGaWn+fBmVlzOcGZWSNFQF892qhOcGaWn2twZtZYTnBm1kgBeGd7M2umgHAfnJk1UeBBBjNrMPfBmVljOcGZWTN5sb2ZNVUAflySmTWWa3Bm1kxeqmVmTRUQngdnZo3llQxm1ljugzOzRorwKKqZNZhrcGbWTEH09ZUdRCZOcGaWjx+XZGaNVpNpIiPKDsDM6iWA6I9Mx85IOkHSXZLulnROt2N1gjOzfCJ94GWWow1JPcAFwGuAg4E3Szq4m6G6iWpmuXVpkOFI4O6IuAdA0uXAbOCObhQOoKjQcK+kdcD9ZcdRgAnA+rKDsFya+nf2vIiYuCsFSLqK5PeTxWjg6ZbzuRExNy3nDcAJEfE36flpwMsi4qxdia9VpWpwu/qLrypJSyNiZtlxWHb+O9uxiDihS0VpsOK7VDbgPjgzK89KYGrL+f7Aqm7ewAnOzMqyBJguaZqk3YBTgIXdvEGlmqgNNrfsACw3/50VLCK2SjoLuBroAS6JiNu7eY9KDTKYmXWTm6hm1lhOcGbWWE5wBSp6GYp1n6RLJK2VdFvZsdiuc4IryFAsQ7FCfAPo1jwvK5kTXHG2LUOJiM3AwDIUq7CIuA7YWHYc1h1OcMWZAjzQcr4yvWZmQ8QJrjiFL0Mxs/ac4IpT+DIUM2vPCa44hS9DMbP2nOAKEhFbgYFlKHcC87u9DMW6T9JlwK+AGZJWSjq97Jisc16qZWaN5RqcmTWWE5yZNZYTnJk1lhOcmTWWE5yZNZYTXI1I6pN0i6TbJF0hacwulPWNdFcjJF3U7kEAkmZJenkH97hP0rN2X9rR9e0+83jOe/2LpA/mjdGazQmuXp6KiBdHxKHAZuDdrW+mTzDJLSL+JiLa7UU5C8id4MzK5gRXX78EXpDWrn4u6VLgt5J6JH1a0hJJt0p6F4ASX5J0h6QfA5MGCpJ0raSZ6esTJC2T9BtJiyUdSJJI35/WHv9M0kRJ30/vsUTSMel3nyNpkaSbJX2VwdfjPoOkH0q6SdLtks7Y7r3PprEsljQxvfZ8SVel3/mlpBd245dpzeRNZ2pI0kiS58xdlV46Ejg0Iu5Nk8QjEfFSSbsD/y1pEXAEMAM4DJhMsnv4JduVOxH4GvCKtKzxEbFR0leAxyPiM+nnLgX+IyKul3QAyWqNFwHnAtdHxMclnQg8I2HtwDvSe+wBLJH0/YjYAOwJLIuID0j6WFr2WSSbwbw7IpZLehlwIfDKDn6NNgw4wdXLHpJuSV//EriYpOl4Y0Tcm15/NfDHA/1rwDhgOvAK4LKI6ANWSbpmkPKPAq4bKCsidvRctFcBB0vbKmhjJe2d3uOv0u/+WNLvM/xMZ0t6ffp6ahrrBqAf+G56/dvAlZL2Sn/eK1ruvXuGe9gw5QRXL09FxItbL6T/0J9ovQS8NyKu3u5zr2Xnj2tShs9A0rVxdEQ8NUgsmdf+SZpFkiyPjognJV0LjN7BxyO978Pb/w7MdsR9cM1zNfAeSaMAJP2RpD2B64BT0j66XuDYQb77K+DPJU1Lvzs+vf4YsHfL5xaRNBdJPzeQcK4DTk2vvQbYdyexjgN+nya3F5LUIAeMAAZqoW8hafo+Ctwr6eT0HpJ0+E7uYcOYE1zzXETSv7Ys3TjlqyQ19R8Ay4HfAl8GfrH9FyNiHUm/2ZWSfsMfmog/Al4/MMgAnA3MTAcx7uAPo7n/CrxC0jKSpvKKncR6FTBS0q3AJ4Bft7z3BHCIpJtI+tg+nl4/FTg9je92/Bh4a8NPEzGzxnINzswaywnOzBrLCc7MGssJzswaywnOzBrLCc7MGssJzswa6/8BUburco4E86YAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from sklearn.metrics import plot_confusion_matrix, classification_report, f1_score\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "m = LogisticRegression(penalty = 'l2', fit_intercept = True)\n",
    "X_train, X_test, y_train, y_test = train_test_split(Xs, y)\n",
    "\n",
    "m.fit(X_train, y_train)\n",
    "print(m.score(X_test, y_test))\n",
    "pred = m.predict(X_test)\n",
    "f1_score(y_test, pred)\n",
    "print(classification_report(y_test, pred))\n",
    "plot_confusion_matrix(m, X_test, y_test)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9700272479564033"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m.score(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "aaa = m.predict(X_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f1_score(aaa, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
